### PR TITLE
[FLINK-34196][Connectors][FLIP-389] Annotate SingleThreadFetcherManager as PublicEvolving.

### PR DESCRIPTION
**Pull request by loserwang1024**
## What is the purpose of the change
As shown in [FLIP-389](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=278465498), this PR has 2 goals:

- To expose the SplitFetcherManager / SingleThreadFetcheManager as Public, allowing connector developers to easily create their own threading models in the SourceReaderBase.

- To hide the element queue from the connector developers and make SplitFetcherManager the only owner class of the queue

## Brief change log

- By exposing the SplitFetcherManager / SingleThreadFetcheManager, connector developers can easily create their own threading models in the SourceReaderBase, by implementing addSplits(), removeSplits() and maybeShutdownFinishedFetchers() functions.
- Note that the SplitFetcher constructor is package private, so users can only create SplitFetchers via SplitFetcherManager.createSplitFetcher(). This ensures each SplitFetcher is always owned by the SplitFetcherManager.

-  This FLIP essentially embedded the element queue (a FutureCompletingBlockingQueue) instance into the SplitFetcherManager. This hides the element queue from the connector developers and simplifies the SourceReaderBase to consist of only SplitFetcherManager and RecordEmitter as major components.

## Verifying this change
This change is already covered by existing tests, such as HybridSourceReaderTest and SplitFetcherManagerTest.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: ( no )
  - The runtime per-record code paths (performance sensitive): ( no )
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? no
  -  how is the feature documented?https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=278465498

### CODE REVIEW

**Comment by loserwang1024** [comment]
@PatrickRen , @leonardBang , CC, WDYT?

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/fetcher/SplitFetcherManager.java [hunk 1]
```diff
@@ -238,6 +274,15 @@ public boolean maybeShutdownFinishedFetchers() {
         return fetchers.isEmpty();
     }
 
+    /**
+     * Return the queue contains data produced by split fetchers.This method is Internal and only
+     * used in {@link SourceReaderBase}.
+     */
+    @Internal
+    public FutureCompletingBlockingQueue getQueue() {
```
Comment (PatrickRen) [inline flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/fetcher/SplitFetcherManager.java:282]: It's better to specify the type parameter of `FutureCompletingBlockingQueue`, which should be `RecordsWithSplitIds<E>`

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/fetcher/SplitFetcherManager.java [hunk 2]
```diff
@@ -151,6 +157,36 @@ public void accept(Throwable t) {
...
+     *
+     * @param splitReaderFactory a supplier that could be used to create split readers.
+     * @param configuration the configuration of this fetcher manager.
+     * @param splitFinishedHook Hook for handling finished splits in split fetchers.
+     */
+    public SplitFetcherManager(
+            Supplier<SplitReader<E, SplitT>> splitReaderFactory,
+            Configuration configuration,
+            Consumer<Collection<String>> splitFinishedHook) {
+        this(
```
Comment (PatrickRen) [inline flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/fetcher/SplitFetcherManager.java:182]: What about using a non-deprecated constructor here?

**Review by PatrickRen — COMMENTED** [review]
@loserwang1024 Thanks for the PR! I left some comments.

**Comment by loserwang1024** [comment]
@flinkbot run azure

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/fetcher/SingleThreadFetcherManager.java [hunk 3]
```diff
@@ -100,6 +104,43 @@ public SingleThreadFetcherManager(
         super(elementsQueue, splitReaderSupplier, configuration, splitFinishedHook);
     }
 
+    /**
+     * Creates a new SplitFetcherManager with a single I/O threads.
+     *
+     * @param splitReaderSupplier The factory for the split reader that connects to the source
+     *     system.
+     */
+    public SingleThreadFetcherManager(Supplier<SplitReader<E, SplitT>> splitReaderSupplier) {
```
Comment (PatrickRen) [inline flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/fetcher/SingleThreadFetcherManager.java:113]: We need to mark the class `SingleThreadFetcherManager` as @PublicEvolving

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/fetcher/SplitFetcherManager.java [hunk 4]
```diff
@@ -151,6 +157,60 @@ public void accept(Throwable t) {
         this.closed = false;
     }
 
+    /**
+     * Create a split fetcher manager.
+     *
+     * @param splitReaderFactory a supplier that could be used to create split readers.
+     * @param configuration the configuration of this fetcher manager.
+     */
+    public SplitFetcherManager(
```
Comment (PatrickRen) [inline flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/fetcher/SplitFetcherManager.java:166]: Also `SplitFetcherManager` should be marked as `@PublicEvolving` now

File: flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/SourceReaderBase.java [hunk 5]
```diff
@@ -136,6 +146,35 @@ public SourceReaderBase(
...
+        this(splitFetcherManager, recordEmitter, null, config, context);
+    }
+
+    public SourceReaderBase(
+            SplitFetcherManager<E, SplitT> splitFetcherManager,
+            RecordEmitter<E, T, SplitStateT> recordEmitter,
+            @Nullable RecordEvaluator<T> eofRecordEvaluator,
+            Configuration config,
+            SourceReaderContext context) {
+        this(
```
Comment (PatrickRen) [inline flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/SourceReaderBase.java:169]: Looks like this is still using a deprecated constructor.

**Review by PatrickRen — COMMENTED** [review]
@loserwang1024 Thanks for the update. I left some comments.

**Comment by loserwang1024** [comment]
@flinkbot run azure

**Pull request merged**